Fix clingo bootstrapping on rhel + ppc64le

alalazo opened the pull request:
fixes #25734 

The system Python interpreter on `rhel` is patched to have slightly different names for some architectures. This makes it incompatible with manylinux generated extensions for `ppc64le`.

To fix this issue when bootstrapping Spack we generate on-the-fly symbolic links to the name expected by the current interpreter if different from the default expectation of CPython.

Links:
-  https://github.com/pypa/manylinux/issues/687
- https://src.fedoraproject.org/fork/churchyard/rpms/python3/blame/00274-fix-arch-names.patch?identifier=test_email-mktime

alalazo committed:
Fix clingo bootstrapping on rhel + ppc64le

The system Python interpreter on rhel is patched to have
slightly different names for some architectures. This
makes it incompatible with manylinux generated extensions
for ppc64le.

To fix this issue when bootstrapping Spack we generate
on-the-fly symbolic links to the name expected by the
current interpreter if it differs from the default.

Links:
  https://github.com/pypa/manylinux/issues/687
  https://src.fedoraproject.org/fork/churchyard/rpms/python3/blame/00274-fix-arch-names.patch?identifier=test_email-mktime

alalazo commented:
@mwkrentel This PR should fix the RHEL 7 issue. Let me know if it works for you

mwkrentel commented:
@alalazo Yes, this did work for me, but I have a couple of questions.

1. I looked inside `bootstrap/store/.../python3.6/site-packages` and I
noticed that the added symlink was a full, absolute path.

```
$ ls -l
total 2752K
drwxr-sr-x. 2 krentel hpc    4096 Apr 12 04:11 clingo
-rwxr-xr-x. 1 krentel hpc 2808408 Sep  2 10:57 clingo.cpython-36m-powerpc64le-linux-gnu.so
lrwxrwxrwx. 1 krentel hpc     196 Sep  2 10:58 clingo.cpython-36m-ppc64le-linux-gnu.so -> /home/krentel/.spack/bootstrap/store/linux-centos7-ppc64le/gcc-9.3.1/clingo-bootstrap-spack-rfu4la457mdbpoffk2g5hikj2hhoek4s/lib/python3.6/site-packages/clingo.cpython-36m-powerpc64le-linux-gnu.so
```

I would think a relative symlink for the same directory would make more sense.

2. I haven't tried this experiment because it takes a while to build,
but instead of the clingo bootstrap, if I built clingo from source,
would clingo + the "broken" RHEL python also work?

3. This is separate from the fix, but when spack installs the
bootstrap, I'd like to see some info message say "installing clingo
bootstrap ..." rather than just a silent huddle for 1.5 minutes.

alalazo commented:
> I haven't tried this experiment because it takes a while to build, but instead of the clingo bootstrap, if I built clingo from source, would clingo + the "broken" RHEL python also work?

I just verified to double check. The build works:
```console
$ /usr/bin/python3 bin/spack find -b
==> Showing internal bootstrap store at "/g/g92/culpo1/.spack/bootstrap/store"
==> 15 installed packages
-- linux-rhel7-ppc64le / gcc@8.3.1 ------------------------------
berkeley-db@18.1.40  clingo-bootstrap@spack  diffutils@3.7  libiconv@1.16  openssl@1.1.1l  pkgconf@1.8.0  re2c@2.2      zlib@1.2.11
bzip2@1.0.8          cmake@3.21.2            gdbm@1.19      ncurses@6.2    perl@5.34.0     python@3.6     readline@8.1

$ /usr/bin/python3 bin/spack solve zlib
==> Best of 2 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                 Value
  1         deprecated versions used                      0
  2         version weight                                0
  3         number of non-default variants (roots)        0
  4         multi-valued variants                         0
  5         preferred providers for roots                 0
  6         number of non-default variants (non-roots)    0
  7         preferred providers (non-roots)               0
  8         compiler mismatches                           0
  9         version badness                               0
  10        count of non-root multi-valued variants       0
  11        non-preferred compilers                       0
  12        target mismatches                             0
  13        non-preferred targets                         0

zlib@1.2.11%gcc@8.3.1+optimize+pic+shared arch=linux-rhel7-power9le
```
This is what I would have expected, since we built with this interpreter and run from this interpreter. The other issue is really because we build the binaries somewhere on a platform that was curated for portability, but the patch to CPython made by RedHat breaks some naming assumptions.

alalazo commented:
> I would think a relative symlink for the same directory would make more sense.

Do you see issues with that or it's just for readability? I'm asking because to be compatible with all the Python versions we support I think I need to do a `os.chdir` dance to get relative paths so I'd be tempted to keep computations (and surface for errors) to the minimum here.

alalazo commented:
> This is separate from the fix, but when spack installs the bootstrap, I'd like to see some info message say "installing clingo bootstrap ..." rather than just a silent huddle for 1.5 minutes.

I'll have a separate PR for this

mwkrentel commented:
> Do you see issues with that or it's just for readability? I'm asking because to be compatible with all the Python versions we support I think I need to do a os.chdir dance to get relative paths so I'd be tempted to keep computations (and surface for errors) to the minimum here.

Mostly cosmetic, but a relative symlink has the advantage that if a
parent directory is moved, then the link still works.  (Plus, it's a
lot easier to inspect visually.)  This is how `libclingo.so` is handled
in `lib64`.

```
$ ls -l
total 8384K
drwxr-sr-x. 3 krentel hpc    4096 Apr 12 04:11 cmake
lrwxrwxrwx. 1 krentel hpc      14 Sep  2 10:57 libclingo.so -> libclingo.so.4
lrwxrwxrwx. 1 krentel hpc      16 Sep  2 10:57 libclingo.so.4 -> libclingo.so.4.0
-rwxr-xr-x. 1 krentel hpc 8577256 Sep  2 10:57 libclingo.so.4.0
```

I guess I'm not understanding the technical problem.
Is it hard to do pushd or chdir() portably in python?

alalazo commented:
> I guess I'm not understanding the technical problem. Is it hard to do pushd or chdir() portably in python?

Not hard, it will be 3 calls instead of one. With Python 3 it's easier, and no need to chdir - but then the API call is not portable to Python 2. I agree on the advantage, but I don't think we'll move the directory around. If we do without relocating other paths the symlink will be the only thing that would continue to work as intended.

mwkrentel commented:
Well, if it's not worth the effort, then whatever. 
I would have guessed that spack would have abstracted that out somewhere.

mwkrentel commented:
@alalazo I apologize for bringing this up.  My mind is trained to look
for unusual, difficult cases.

What if I use the "broken" RHEL python to build clingo from source and
then later I build and start using a later python.  Will clingo work
with the later, "correct" python?

Maybe it would be safer to always install both names for ppc64.

alalazo commented:
> I apologize for bringing this up. My mind is trained to look for unusual, difficult cases.

That's good, I prefer to think about those as soon as possible. It's what usually complicates an otherwise simple design.

> What if I use the "broken" RHEL python to build clingo from source and then later I build and start using a later python. Will clingo work with the later, "correct" python?

I expect it would work, but I'll double check. During bootstrapping Spack transitorily swaps configuration and registers the Python interpreter being used as an external Python, thus that Python will enter the hash. The bootstrapping process in the case you mentioned should do the following:
1. Try the clingo installed with RHEL Python (this should fail)
2. Concretize a new clingo spec and try to install it (this should be different from the one before, due to interpreter path)
3. Use clingo at point 2

alalazo commented:
So, I tried different combinations of bootstrapping from sources and they all worked, except one:
1. Use the system Python interpreter to bootstrap clingo
2. Use a non-patched interpreter **at the same minor version** (so in our case Python 3.6) to run Spack later

_In principle in should work_ :tm: but since `clingo` installs a directory named "clingo" with just type information in it, Python 3 namespace packages protocol kicks in at lowest priority and loads an empty Python module instead of failing. Installing clingo in the opposite way (non-patched interpreter first, system Python later) works instead, as per the fix in this PR.

alalazo committed:
Account for bootstrapping from sources niche case

This modification accounts for:
1. Bootstrapping from sources using system, non-standard Python
2. Using later an ABI compatible standard Python interpreter

mwkrentel commented:
@alalazo I'm not sure I understand the multiple python case.  If I build clingo
from source with the "broken" RHEL python, the build a newer python,
then use the new python, are you saying that the new python will fail
with the old clingo and then automatically rebuild clingo?
I don't know if I understand that correctly.

If that is so, then wouldn't it be better to just always install two
names for clingo.cpython on powerpc for both bootstrap and build from
source?

alalazo commented:
> I'm not sure I understand the multiple python case. 

I'll try to simplify. Python is ABI incompatible across minor versions, so if you use rhel system Python 3.6 to build e.g. a Python 3.9 you'll need to bootstrap clingo again for Python 3.9 later (due to the ABI incompatibility).

If you build instead another Python 3.6 6169485ce7809f3ed6fc6fcd3f392a1cc8c6d8f2 ensures that things are linked with both naming conventions. So currently this should work out of the box and avoid unnecessary rebuilds.

Note: I thought the second case would have worked without the last commit but I didn't take into account a Python 3 "feature", which is support for "namespace packages". Since it happens that `clingo` installs in the same directory where the extension is placed a `clingo` subdirectory with metadata on types, Python thinks it's an empty `clingo` module when it is not able to load the shared library.

mwkrentel commented:
Ok, now I understand.  I was fooled by the compatibility of the bootstrap binary to think that that one clingo should work for all possible python.  Never mind.
